server-side/framework/python/django/lessons-learned-from-ultra-gro-deployment.md

```diff
@@ -0,0 +1,17 @@
+# Lessons Learned from Ultra-Gro Deployment
+
+This project was a Dockerized Django web app using a Postgres variant called PostGIS for the database server. We deployed it to an AWS EC2 instance running Ubuntu 18.04 LTS, with nginx as the front end.
+
+## Gotchas to look out for
+
+`nginx` misconfiguration. Read the nginx [docs](https://www.nginx.com/resources/wiki/start/topics/tutorials/config_pitfalls/) for common config mistakes, including some made by `cert-bot` during LetsEncrypt setup. In particular, nginx highly discourages using `if` blocks.
+
+Firewall issues. Be sure to open the appropriate ports through AWS’ / Azure’s firewall. (SSH, HTTP, HTTPS, possibly others.) Also, I recommend that you use an internal firewall, such as `ufw`. (`ufw` is an Ubuntu firewall tool — actually a front end for `iptables` — which allows you to lock down all ports except the ones that really need to be open.) Be sure to open the appropriate ports through that as well — and to allow communication within Docker’s internal network(s), if needed.
+
+`gunicorn` configuration issues. In the Ultra-Gro site, gunicorn had a host whitelist built into the .py settings file. I had to add the external domain name to it ([ultra-gro.shift3sandbox.com](https://ultra-gro.shift3sandbox.com), in this case). I was able to do this using environment settings.
```

Comment:
We should explicitly mention "ALLOWED_HOSTS" of django settings.

Reply:
Done.